superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx

```diff
@@ -0,0 +1,602 @@
+      overrides: { since: null, until: null },
+    });
+  }
+
+  fetchOptions(annotationType, sourceType, isLoadingOptions) {
+    if (isLoadingOptions === true) {
+      if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
+        $.ajax({
+          type: 'GET',
+          url: '/annotationlayermodelview/api/read?',
+        }).then((data) => {
+          const layers = data ? data.result.map(layer => ({
+            value: layer.id,
+            label: layer.name,
+          })) : [];
+          this.setState({
+            isLoadingOptions: false,
+            valueOptions: layers,
+          });
+        });
+      } else if (requiresQuery(sourceType)) {
+        $.ajax({
+          type: 'GET',
+          url: '/superset/user_slices',
+        }).then(data =>
+          this.setState({
+            isLoadingOptions: false,
+            valueOptions: data.filter(
+                x => getSupportedSourceTypes(annotationType)
+                .find(v => v === x.viz_type))
+                .map(x => ({ value: x.id, label: x.title, slice: x }),
+              ),
+          }),
+        );
+      } else {
+        this.setState({
+          isLoadingOptions: false,
+          valueOptions: [],
+        });
+      }
+    }
+  }
+
+  deleteAnnotation() {
+    this.props.close();
+    if (!this.state.isNew) {
+      this.props.removeAnnotationLayer(this.state);
+    }
+  }
+
+  applyAnnotation() {
+    if (this.state.name.length) {
+      const annotation = { ...this.state };
+      annotation.color = annotation.color === AUTOMATIC_COLOR ? null : annotation.color;
+      delete annotation.isNew;
+      delete annotation.valueOptions;
+      delete annotation.isLoadingOptions;
+      this.props.addAnnotationLayer(annotation);
+      this.setState({ isNew: false, oldName: this.state.name });
+    }
+  }
+
+  submitAnnotation() {
+    this.applyAnnotation();
+    this.props.close();
+  }
+
+  renderValueConfiguration() {
+    const { annotationType, sourceType, value,
+      valueOptions, isLoadingOptions } = this.state;
+    let label = '';
+    let description = '';
+    if (requiresQuery(sourceType)) {
+      if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
+        label = 'Annotation Layer';
+        description = 'Select the Annotation Layer you would like to use.';
+      } else {
+        label = 'Slice';
+        description = `Use a pre defined Superset Slice as a source for annotations and overlays. 
+        'your Slice must be one of these visualization types:
+        '[${getSupportedSourceTypes(sourceType)
+            .map(x => vizTypes[x].label).join(', ')}]'`;
+      }
+    } else if (annotationType === AnnotationTypes.FORMULA) {
+      label = 'Formula';
+      description = `Expects a formula with depending time parameter 'x'
+        in milliseconds since epoch. mathjs is used to evaluate the formulas.
+        Example: '2x+5'`;
+    }
+    if (requiresQuery(sourceType)) {
+      return (
+        <SelectControl
```

Comment:
You may want to use the `SelectAsyncControl` control here that works well with the FAB CRUD REST api and manages its own state.

Reply:
I tried to used that but ran into issues with it. I can't remember what they were since I wrote this 4 month ago.